File: android/jsbridge/src/main/java/net/kibotu/jsbridge/JavaScriptBridge.kt
package net.kibotu.jsbridge

import android.webkit.JavascriptInterface
import android.webkit.WebView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import net.kibotu.jsbridge.commands.BridgeCommand
import net.kibotu.jsbridge.commands.utils.BridgeResponseUtils
import net.kibotu.jsbridge.decorators.BridgeWebViewClient
import org.json.JSONObject
import timber.log.Timber
import java.util.WeakHashMap

/**
 * Enables seamless bidirectional communication between web content and native Android.
 *
 * Web content interacts via `window.<bridgeName>` (default: `window.jsbridge`).
 * The bridge name is configurable so multiple bridges can coexist on the same WebView,
 * each with its own set of commands.
 *
 * @param webView The WebView instance to bridge with
 * @param commands List of command handlers this bridge responds to
 * @param bridgeName Name exposed to JavaScript (default [DEFAULT_BRIDGE_NAME])
 * @see <a href="https://github.com/kibotu/jsbridge">jsbridge specification</a>
 */
class JavaScriptBridge(
    private val webView: WebView,
    private val commands: List<BridgeCommand>,
    private val bridgeName: String = DEFAULT_BRIDGE_NAME
) {
    /** The WebView's context, used by commands and services to resolve Activity/Application. */
    val context get() = webView.context

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val callbackResponse get() = "__${bridgeName}_handleResponse"
    private val callbackMessage get() = "__${bridgeName}_handleNativeMessage"

    companion object {
        const val DEFAULT_BRIDGE_NAME = "jsbridge"
        const val SCHEMA_VERSION = 1

        private val bridges = WeakHashMap<WebView, MutableMap<String, JavaScriptBridge>>()

        /**
         * Retrieve the bridge associated with a WebView by name.
         * Falls back to the default bridge name if not specified.
         */
        fun WebView.bridge(name: String = DEFAULT_BRIDGE_NAME): JavaScriptBridge? =
            bridges[this]?.get(name)

        /**
         * One-liner to set up the JavaScript bridge on a [WebView].
         *
         * - Creates the bridge and registers the `@JavascriptInterface`
         * - Wraps the existing [android.webkit.WebViewClient] with [BridgeWebViewClient]
         *   (decorator pattern) so bridge injection and safe area CSS happen automatically
         * - Associates the bridge with the WebView for retrieval via [bridge]
         *
         * ```kotlin
         * val bridge = JavaScriptBridge.inject(
         *     webView = webView,
         *     commands = listOf(DeviceInfoCommand(), ShowToastCommand()),
         *     bridgeName = "jsbridge"
         * )
         * ```
         *
         * @return The created [JavaScriptBridge] instance.
         */
        fun inject(
            webView: WebView,
            commands: List<BridgeCommand>,
            bridgeName: String = DEFAULT_BRIDGE_NAME
        ): JavaScriptBridge {
            val bridge = JavaScriptBridge(webView, commands, bridgeName)
            webView.addJavascriptInterface(bridge, bridgeName)

            val map = bridges.getOrPut(webView) { mutableMapOf() }
            map[bridgeName] = bridge

            val currentClient = webView.webViewClient
            if (currentClient !is BridgeWebViewClient) {
                webView.webViewClient = BridgeWebViewClient(currentClient)
            }

            Timber.d("[Bridge] Injected (name=$bridgeName, commands=${commands.size})")
            return bridge
        }
    }

    /**
     * Entry point for all web-to-native communication.
     *
     * @param message JSON string from web: `{ id?, version?, data: { action, content? } }`
     */
    @Suppress("unused")
    @JavascriptInterface
    fun postMessage(message: String) {
        Timber.d("[postMessage] received: $message")

        scope.launch {
            var id: String? = null
            try {
                val messageObj = JSONObject(message)
                id = messageObj.optString("id", null)
                val data = messageObj.optJSONObject("data")

                if (data == null) {
                    sendErrorToWeb(id, "INVALID_MESSAGE", "Missing 'data' field in message")
                    return@launch
                }

                val requestedVersion = messageObj.optInt("version", SCHEMA_VERSION)
                if (requestedVersion > SCHEMA_VERSION) {
                    Timber.w("[postMessage] Ignoring message with version $requestedVersion (current: $SCHEMA_VERSION)")
                    return@launch
                }

                val action = data.optString("action", null)
                if (action.isNullOrEmpty()) {
                    sendErrorToWeb(id, "INVALID_MESSAGE", "Missing 'action' field in data")
                    return@launch
                }

                val content = data.opt("content")
                val command = commands.find { it.action == action }

                if (command == null) {
                    Timber.w("[postMessage] Unknown action: $action")
                    sendErrorToWeb(id, "UNKNOWN_ACTION", "Unknown action: $action")
                    return@launch
                }

                val result = command.handle(content)

                if (!id.isNullOrEmpty()) {
                    sendResponseToWeb(id, result)
                }
            } catch (e: Exception) {
                Timber.e(e)
                sendErrorToWeb(id, "INTERNAL_ERROR", e.message ?: "Unknown error")
            }
        }
    }

    /**
     * Cancels all in-flight coroutines and removes the bridge association.
     * Call when the hosting component (Fragment, Activity) is destroyed.
     */
    fun destroy() {
        scope.cancel()
        bridges[webView]?.remove(bridgeName)
        if (bridges[webView]?.isEmpty() == true) {
            bridges.remove(webView)
        }
        Timber.d("[Bridge] Destroyed (name=$bridgeName)")
    }

    /**
     * Send a message from native to web (push notifications, state updates, events).
     */
    fun sendToWeb(action: String, content: Any? = null) {
        scope.launch {
            try {
                val message = buildNativeMessage(action, content)
                val script = "window.$callbackMessage && window.$callbackMessage($message)"
                executeJavaScript(script)
                Timber.d("[sendToWeb] action=$action content=$content")
            } catch (e: Exception) {
                Timber.e(e)
            }
        }
    }

    private fun sendResponseToWeb(id: String, result: Any?) {
        if (result is JSONObject && result.has("error")) {
            val error = result.optJSONObject("error")
            if (error != null) {
                sendErrorToWeb(id, error.optString("code", "UNKNOWN"), error.optString("message", "Unknown error"))
                return
            }
        }

        val response = JSONObject().apply {
            put("id", id)
            put("data", result)
        }

        val script = "window.$callbackResponse && window.$callbackResponse($response)"
        executeJavaScript(script)

        Timber.d("[sendResponseToWeb] id=$id result=$result")
    }

    private fun sendErrorToWeb(id: String?, code: String, message: String) {
        val error = JSONObject().apply {
            put("code", code)
            put("message", message)
        }

        val response = JSONObject().apply {
            if (!id.isNullOrEmpty()) put("id", id)
            put("error", error)
        }

        val script = "window.$callbackResponse && window.$callbackResponse($response)"
        executeJavaScript(script)

        Timber.w("[sendErrorToWeb] id=$id code=$code message=$message")
    }

    private fun buildNativeMessage(action: String, content: Any?): String {
        val message = JSONObject()
        val data = JSONObject().apply {
            put("action", action)
            if (content != null) {
                val jsonContent = when (content) {
                    is Map<*, *> -> JSONObject(content)
                    else -> content
                }
                put("content", jsonContent)
            }
        }
        message.put("data", data)
        return message.toString()
    }

    private fun executeJavaScript(script: String) {
        webView.post {
            webView.evaluateJavascript(script, null)
        }
    }

    /**
     * Inject safe area CSS custom properties into the WebView.
     * Call this whenever bars change visibility, on rotation, or on keyboard show/hide.
     */
    fun updateSafeAreaCSS(
        insetTop: Int = 0,
        insetBottom: Int = 0,
        insetLeft: Int = 0,
        insetRight: Int = 0,
        statusBarHeight: Int = 0,
        topNavHeight: Int = 0,
        bottomNavHeight: Int = 0,
        systemNavHeight: Int = 0
    ) {
        val js = """
            (function() {
                var el = document.documentElement;
                el.style.setProperty('--bridge-inset-top', '${insetTop}px');
                el.style.setProperty('--bridge-inset-bottom', '${insetBottom}px');
                el.style.setProperty('--bridge-inset-left', '${insetLeft}px');
                el.style.setProperty('--bridge-inset-right', '${insetRight}px');
                el.style.setProperty('--bridge-status-bar', '${statusBarHeight}px');
                el.style.setProperty('--bridge-top-nav', '${topNavHeight}px');
                el.style.setProperty('--bridge-bottom-nav', '${bottomNavHeight}px');
                el.style.setProperty('--bridge-system-nav', '${systemNavHeight}px');
            })();
        """.trimIndent()
        executeJavaScript(js)
    }

    /**
     * Injects the unified bridge JavaScript API into the WebView.
     * Reads from res/raw/bridge.js and replaces template variables.
     * Call on page load and after navigation.
     */
    fun injectBridgeScript() {
        val script = loadBridgeScript()
        executeJavaScript(script)
        Timber.d("[injectBridgeScript] Bridge script injected (bridgeName=$bridgeName)")
    }

    private fun loadBridgeScript(): String {
        val rawScript = webView.context.resources
            .openRawResource(R.raw.bridge)
            .bufferedReader()
            .use { it.readText() }

        return rawScript
            .replace("__BRIDGE_NAME__", bridgeName)
            .replace("__SCHEMA_VERSION__", SCHEMA_VERSION.toString())
    }
}

File: ios/JSBridge/Sources/JSBridge/JavaScriptBridge.swift
import Foundation
import WebKit
import Orchard

/// Weak proxy to avoid the WKUserContentController retain cycle.
///
/// `WKUserContentController.add(_:name:)` retains its delegate strongly.
/// Without this proxy, the bridge and the WebView's content controller form a retain cycle:
/// contentController -> bridge -> (weak) webView -> configuration -> contentController.
@MainActor
private final class LeakAvoider: NSObject, WKScriptMessageHandler {
    weak var delegate: (any WKScriptMessageHandler)?

    init(delegate: any WKScriptMessageHandler) {
        self.delegate = delegate
        super.init()
    }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        delegate?.userContentController(userContentController, didReceive: message)
    }
}

/// Main JavaScript bridge coordinator for WKWebView <-> native communication.
///
/// Configurable bridge name (default: `jsbridge`). The same name is used for:
/// - The WKScriptMessageHandler registration
/// - The `window.<name>` public JS API
/// - The global callback names (`window.__<name>_handleResponse`, etc.)
///
/// Commands are passed in at init time -- they are **not** auto-registered.
/// Use `DefaultCommands.all(...)` for a convenient all-in-one setup.
///
/// Multiple bridges can coexist on the same WKWebView as long as they
/// use different names.
@MainActor
public final class JavaScriptBridge: NSObject, WKScriptMessageHandler {
    private var commands: [any BridgeCommand] = []
    private weak var webView: WKWebView?
    private weak var viewController: UIViewController?
    private var bridgeScriptInjected = false

    public let name: String

    private let schemaVersion: Int = 1

    private var callbackResponse: String { "__\(name)_handleResponse" }
    private var callbackMessage: String { "__\(name)_handleNativeMessage" }

    public static let defaultBridgeName = "jsbridge"

    /// Creates and configures the bridge.
    ///
    /// - Parameters:
    ///   - webView: The WKWebView to bridge with.
    ///   - viewController: The hosting view controller (used by UI commands like alerts/toasts).
    ///   - bridgeName: The name exposed to JavaScript (default: `jsbridge`).
    ///   - commands: The commands this bridge responds to.
    public init(
        webView: WKWebView,
        viewController: UIViewController,
        bridgeName: String = defaultBridgeName,
        commands: [any BridgeCommand]
    ) {
        self.name = bridgeName
        self.webView = webView
        self.viewController = viewController
        super.init()

        for command in commands {
            register(command: command)
        }

        setupMessageHandler()
        injectBridgeScript()
    }

    deinit {
        let name = self.name
        let controller = webView?.configuration.userContentController
        MainActor.assumeIsolated {
            controller?.removeScriptMessageHandler(forName: name)
        }
    }

    // MARK: - Setup

    private func setupMessageHandler() {
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: name)
        webView?.configuration.userContentController.add(LeakAvoider(delegate: self), name: name)
    }

    private func register(command: any BridgeCommand) {
        commands.append(command)
        Orchard.v("[Bridge] Registered command for action: \(command.action)")
    }

    private func command(for action: String) -> (any BridgeCommand)? {
        return commands.first { $0.action == action }
    }

    nonisolated private func isVersionSupported(_ version: Int) -> Bool {
        return version <= schemaVersion
    }

    /// Injects the unified bridge JavaScript into the WebView at document start.
    /// Reads bridge.js from the bundle resource and replaces template variables.
    /// Guards against duplicate injection to prevent WKUserScript accumulation.
    public func injectBridgeScript() {
        guard !bridgeScriptInjected else {
            Orchard.v("[Bridge] Bridge script already injected, skipping")
            return
        }

        let script = loadBridgeScript()
        let userScript = WKUserScript(
            source: script,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )

        webView?.configuration.userContentController.addUserScript(userScript)
        bridgeScriptInjected = true
        Orchard.v("[Bridge] Bridge script injected (name=\(name), schema v\(schemaVersion))")
    }

    nonisolated private func loadBridgeScript() -> String {
        guard let url = Bundle.module.url(forResource: "bridge", withExtension: "js"),
              let rawScript = try? String(contentsOf: url, encoding: .utf8) else {
            Orchard.e("[Bridge] Failed to load bridge.js from bundle")
            return ""
        }

        return rawScript
            .replacingOccurrences(of: "__BRIDGE_NAME__", with: name)
            .replacingOccurrences(of: "__SCHEMA_VERSION__", with: String(schemaVersion))
    }

    // MARK: - WKScriptMessageHandler

    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == name else { return }

        guard let body = message.body as? String else {
            Orchard.w("[Bridge] Received invalid message body")
            return
        }

        handleMessage(body)
    }

    // MARK: - Message Handling

    private func handleMessage(_ messageString: String) {
        guard let data = messageString.data(using: .utf8) else {
            Orchard.e("[Bridge] Could not convert message to data")
            return
        }

        let decoder = JSONDecoder()

        guard let message = try? decoder.decode(JavaScriptBridgeMessage.self, from: data) else {
            Orchard.e("[Bridge] Could not decode message: \(messageString)")
            let rawId = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["id"] as? String
            if let rawId = rawId {
                sendError(id: rawId, error: .invalidMessage)
            }
            return
        }

        if !isVersionSupported(message.version) {
            Orchard.w("[Bridge] Silently ignoring unsupported version: \(message.version)")
            return
        }

        let action = message.data.action
        let content = message.data.content?.compactMapValues { $0.value is NSNull ? nil : $0.value }

        Orchard.v("[Bridge] Received action: \(action)")

        guard let command = command(for: action) else {
            Orchard.w("[Bridge] Unknown action: \(action)")
            sendError(id: message.id, error: .unknownAction(action))
            return
        }

        let messageId = message.id
        Task { [weak self] in
            do {
                let responseData = try await command.handle(content: content)
                self?.sendSuccess(id: messageId, data: responseData)
            } catch let error as BridgeError {
                self?.sendError(id: messageId, error: error)
            } catch {
                self?.sendError(id: messageId, error: .internalError(error.localizedDescription))
            }
        }
    }

    // MARK: - Sending Responses (unified format: { id, data?, error? })

    private func sendSuccess(id: String, data: [String: Any]?) {
        var response: [String: Any] = ["id": id]
        if let data = data {
            response["data"] = data
        } else {
            response["data"] = [String: Any]()
        }
        sendResponseJSON(response)
    }

    private func sendError(id: String, error: BridgeError) {
        let response: [String: Any] = [
            "id": id,
            "error": [
                "code": error.code,
                "message": error.message
            ]
        ]
        sendResponseJSON(response)
    }

    private func sendResponseJSON(_ response: [String: Any]) {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: response),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            Orchard.e("[Bridge] Failed to encode response")
            return
        }

        let script = "window.\(callbackResponse) && window.\(callbackResponse)(\(jsonString));"

        webView?.evaluateJavaScript(script) { _, error in
            if let error = error {
                Orchard.e("[Bridge] Failed to send response: \(error)")
            }
        }
    }

    // MARK: - Native to Web Communication

    /// Send a message from native to web (fire-and-forget).
    public func sendToWeb(action: String, content: [String: Any]? = nil) {
        let message: [String: Any] = [
            "data": [
                "action": action,
                "content": content ?? [:]
            ]
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: message),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            Orchard.e("[Bridge] Failed to serialize message to web")
            return
        }

        let script = "window.\(callbackMessage) && window.\(callbackMessage)(\(jsonString));"

        webView?.evaluateJavaScript(script) { _, error in
            if let error = error {
                Orchard.e("[Bridge] Failed to send message to web: \(error)")
            }
        }
    }

    // MARK: - Safe Area CSS Injection

    /// Injects/updates CSS custom properties for safe area insets.
    /// Call whenever bars change visibility, on rotation, or keyboard show/hide.
    public func updateSafeAreaCSS(
        insetTop: CGFloat = 0,
        insetBottom: CGFloat = 0,
        insetLeft: CGFloat = 0,
        insetRight: CGFloat = 0,
        statusBarHeight: CGFloat = 0,
        topNavHeight: CGFloat = 0,
        bottomNavHeight: CGFloat = 0,
        systemNavHeight: CGFloat = 0
    ) {
        let js = """
        (function() {
            var el = document.documentElement;
            el.style.setProperty('--bridge-inset-top', '\(Int(insetTop))px');
            el.style.setProperty('--bridge-inset-bottom', '\(Int(insetBottom))px');
            el.style.setProperty('--bridge-inset-left', '\(Int(insetLeft))px');
            el.style.setProperty('--bridge-inset-right', '\(Int(insetRight))px');
            el.style.setProperty('--bridge-status-bar', '\(Int(statusBarHeight))px');
            el.style.setProperty('--bridge-top-nav', '\(Int(topNavHeight))px');
            el.style.setProperty('--bridge-bottom-nav', '\(Int(bottomNavHeight))px');
            el.style.setProperty('--bridge-system-nav', '\(Int(systemNavHeight))px');
        })();
        """

        webView?.evaluateJavaScript(js) { _, error in
            if let error = error {
                Orchard.e("[Bridge] Failed to inject safe area CSS: \(error)")
            }
        }
    }

    // MARK: - Lifecycle Events

    public func notifyLifecycleEvent(_ event: String) {
        sendToWeb(action: "lifecycle", content: ["event": event])
    }
}
